src/common/pagination.py

```diff
@@ -148,13 +159,16 @@ def decode_pagination_token(
 
 
 def encode_pagination_token(
-    pagination_token: str, encrypted: bool, encryption_client: EncryptionClient = None
+    pagination_token: str,
+    encrypted: bool,
+    encryption_client: EncryptionClient = None,
+    encryption_context: Union[None, dict] = None,
```

Comment:
The functions that know they are encrypting pagination tokens should take the desired context directly, e.g., user and path, rather than an arbitrary, untyped context.

Reply:
That's fair. In this case I'd suggest including the API Gateway API ID in the context as this deep in the utility there's no guarantee that the caller knows about the convention "don't share encryption material across API Gateway instances". Thoughts? 

Related: would you suggest creating an opinionated `PaginationEncryptionContext` class that can be instantiated using an event and has an `as_dict` method to satisfy the encryption SDK? What's the Python naming convention here -- I feel like `pagination.PaginationEncryptionContext` doesn't scan well, but `import EncryptionContext from common.pagination` loses meaning.

Reply:
Is it true, though, that we don't want to share encryption material across API GW instances? Are there situations where that might be justified?

I would say I'd prefer `encode_pagination_token(pagination_token, api_id, user_id, path)` to `encode_pagination_token(pagination_token, context)`

Reply:
That's because these functions are defined in, and dedicated to, this service. If it was, instead, a general-purpose library for encoding pagination tokens (something that may be worth doing!) a flexible Context with helper functions to extract from an event, etc. would probably make sense.

Reply:
> Is it true, though, that we don't want to share encryption material across API GW instances? Are there situations where that might be justified?

Different teams work in different ways. My thought process here was that if you don't share encryption material, then key + user + path is likely sufficient to ensure that tokens can't be re-used in an unintended context. If you do share encryption material, then key + user + APIGW ID + path may be needed to be positive that you don't have collisions (though it's pretty unlikely that path would be the same across two APIGW's that share key material).

> I would say I'd prefer `encode_pagination_token(pagination_token, api_id, user_id, path)` to `encode_pagination_token(pagination_token, context)`

I've developed an allergy to >2 positional parameters of the same type in a row; while in this context we're likely to always use variables, I've seen too many calls that look like `fn("a", "b", "c", "d")` and struggled to understand what each parameter was for. I'd be far more comfortable with either a class that can extract from an event or named parameters. At the end of the day, though, it's your repository I'm writing a PR for. 🙂 

Thoughts on building a proto-Context here with the idea that it could be pulled out along with the rest of `pagination.py` to a general-purpose library if that became interesting?

Reply:
> Different teams work in different ways.

Which would we rather recommend: "use different keys" or "use different context"? Like if we had to pick. I'm not sure but I don't think the answer is recommending "use different keys AND different context", but maybe it is.

> I've developed an allergy to >2 positional parameters of the same type in a row

This is absolutely correct, and I noticed in this review that I had not made these arguments keyword-only. I mean we should define it as something like `encode_pagination_token(pagination_token, *, path, user=None)`

> Thoughts on building a proto-Context here

I'd rather do it separately. One option is to build that separate library now and then have this project use it.

Reply:
> Which would we rather recommend: "use different keys" or "use different context"? Like if we had to pick. I'm not sure but I don't think the answer is recommending "use different keys AND different context", but maybe it is.

My default would be "use different keys". The teams I work with deploy their keys as part of their service stack; each service has its own APIGW. It is completely and utterly coincidental (sincerely) that this means they wouldn't need the APIGW in the context.
 
> > I've developed an allergy to >2 positional parameters of the same type in a row
> 
> This is absolutely correct, and I noticed in this review that I had not made these arguments keyword-only. I mean we should define it as something like `encode_pagination_token(pagination_token, *, path, user=None)`
> > Thoughts on building a proto-Context here
> 
> I'd rather do it separately. One option is to build that separate library now and then have this project use it.

Whatever works for you. If you don't want to sidetrack, I'm happy to change this PR to use a bunch of keyword parameters.